src/Microsoft.Azure.SignalR.Common/ServiceConnections/scopes/AsyncLocalScopeBase.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal abstract class AsyncLocalScopeBase<TProps> : IDisposable
```

Comment:
>AsyncLocalScopeBase [](start = 28, length = 19)

who else uses this base class?

Reply:
We have `ClientConnectionScope`, very similar to `ServiceConnectioncontainerScope`, I want to make them inherit this base class later, but in another PR.

Reply:
consider reuse the same pattern of `ClientConnectionScope` to make the usage instead of adding more classes?

---
In reply to: [450011381](https://github.com/Azure/azure-signalr/pull/923#discussion_r450011381) [](ancestors = 450011381)

Reply:
> consider reuse the same pattern of `ClientConnectionScope` to make the usage simple instead of adding more classes?
> 

Try to protect the `IServiceDiagnosticLogsContext` inside asyncLocal. The `ScopePropertiesAccessor` is internal, can be accessed everywhere. I don't want the `IServiceDiagnosticLogsContext` accidently modified. The only right place to update the `IServiceDiagnosticLogsContext` is `ServiceConnectionContainerBase.HandlePingAsync()`, and no plan to modify else where. So I think it's better to limit the ability to modify `IServiceDiagnosticLogsContext`.
I updated the some [code](https://github.com/Azure/azure-signalr/pull/923/files#diff-d4a73abdc56033c8cc3e130bd06d079dR12) and [test](https://github.com/Azure/azure-signalr/pull/923/files#diff-bf96c03afebc543419254b1a6fd1baf6R1538).

Reply:
I mean reuse `ScopePropertiesAccessor` in the `ServiceConnectionContainerScope`, any concerns in reusing that?

---
In reply to: [450045026](https://github.com/Azure/azure-signalr/pull/923#discussion_r450045026) [](ancestors = 450045026)

Reply:
make sense! The `ClientConnectionScope` can limit the acess since the `ClientConnectionScopeProperties` a private class. But I can't since I need to add a member _serviceDiagnosticLogsContext [here](https://github.com/Azure/azure-signalr/pull/923/files#diff-72c9ae4e6a93ede091bc3b08bae82afaR46) to avoid using event and event handlers. So `ServiceDiagnosticLogsContext` should be internal, everywhere can use ServiceDiagnosticLogsContext, including `ScopePropertiesAccessor<IServiceDiagnosticLogsContext>` every where.

Reply:
how about exposing a method in `ServiceConnectionContainerScope` to `EnableMessageLog` instead of having the context outside of the scope?

---
In reply to: [450049529](https://github.com/Azure/azure-signalr/pull/923#discussion_r450049529) [](ancestors = 450049529,450045026)